tutorials/detailed_model_reference.ipynb

```diff
@@ -0,0 +1,663 @@
+     "start_time": "2024-07-12T12:16:43.449203Z"
+    }
+   },
+   "source": [
+    "class GeometricAlgebraInterface:\n",
+    "    num_input_channels = num_output_channels = 1\n",
+    "    num_input_scalars = num_output_scalars = 1\n",
+    "\n",
+    "    @staticmethod\n",
+    "    @torch.no_grad()\n",
+    "    def embed(data):\n",
+    "\n",
+    "        multivectors = embed_oriented_plane(normal=data.orientation, position=data.pos).view(-1, 1, 16)\n",
+    "        scalars = data.scalar_feature.view(-1, 1)\n",
+    "\n",
+    "        return multivectors, scalars\n",
+    "\n",
+    "    @staticmethod\n",
+    "    def dislodge(multivectors, scalars):\n",
+    "        return extract_translation(multivectors).squeeze()\n"
+   ],
+   "outputs": [],
+   "execution_count": 5
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c832f5754fbd5911",
+   "metadata": {},
+   "source": [
+    "### 2.1 Tokenizer\n",
+    "```TODO```"
+   ]
+  },
+  {
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-07-12T12:16:44.050245Z",
+     "start_time": "2024-07-12T12:16:44.045916Z"
+    }
+   },
+   "cell_type": "code",
+   "source": [
+    "def interp(\n",
+    "    mlp: torch.nn.Module,\n",
+    "    multivectors: torch.Tensor,\n",
+    "    multivectors_skip: torch.Tensor,\n",
+    "    scalars: torch.Tensor,\n",
+    "    scalars_skip: torch.Tensor,\n",
+    "    pos_source: torch.Tensor,\n",
+    "    pos_target: torch.Tensor,\n",
+    "    data: Data,\n",
+    "    scale_id: int,\n",
+    "    reference_multivector: torch.Tensor\n",
+    ") -> torch.Tensor:\n",
+    "\n",
+    "    pos_diff = pos_source[data[f'scale{scale_id}_interp_source']] - pos_target[data[f'scale{scale_id}_interp_target']]\n",
+    "    squared_pos_dist = torch.clamp(torch.sum(pos_diff ** 2, dim=-1), min=1e-16).view(-1, 1, 1)\n",
+    "\n",
+    "    denominator = scatter(1. / squared_pos_dist, data[f'scale{scale_id}_interp_target'].long(), dim=0, reduce='sum')\n",
+    "\n",
+    "    multivectors = scatter(\n",
+    "        multivectors[data[f'scale{scale_id}_interp_source']] / squared_pos_dist,\n",
+    "        data[f'scale{scale_id}_interp_target'].long(),\n",
+    "        dim=0,\n",
+    "        reduce='sum'\n",
+    "    ) / denominator\n",
+    "\n",
+    "    scalars = scatter(\n",
+    "        scalars[data[f'scale{scale_id}_interp_source']] / squared_pos_dist.view(-1, 1),\n",
+    "        data[f'scale{scale_id}_interp_target'].long(),\n",
+    "        dim=0,\n",
+    "        reduce='sum'\n",
+    "    ) / denominator.view(-1, 1)\n",
+    "\n",
+    "    multivectors = torch.cat((multivectors, multivectors_skip), dim=-2)\n",
+    "    scalars = torch.cat((scalars, scalars_skip), dim=-1)\n",
+    "\n",
+    "    return mlp(multivectors, scalars, reference_mv=reference_multivector)"
+   ],
+   "id": "692225270547a492",
+   "outputs": [],
+   "execution_count": 6
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "```TODO```",
```

Comment:
The pooling layer is implemented via PyG message passing. We just have to embed the relative position between points as multivector.